### PR TITLE
Bookmarks

### PR DESCRIPTION
**Pull request by yimingnzhao**
Bookmark functionality for user personalization. Includes the following:
- Bookmark button to add bookmarks to video via form
- Bootstrap cards to display bookmarks for the current video and user
- 'Show' button to show the contents of the button and jumps to the timestamp of the bookmark
- 'Delete' button to remove the bookmark from the database
- `?mockall` will now also mock the current user, with the common email 'MOCK'

There's also some servlet tests, database interfaces/implementations, and helper classes included with this pull request.

The code is deployed here:
[https://yimingnzhao-dot-step-intern-2020.wl.r.appspot.com/](https://yimingnzhao-dot-step-intern-2020.wl.r.appspot.com/)

### CODE REVIEW

File: src/main/java/com/google/sps/storage/BookmarkStorageManager.java [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+    /**
+     * Removes a bookmark from the database
+     * @param id The globally unique id of the bookmark to remove
+     */
+    public void removeBookmark(String id) throws BookmarkStorageException {
+        // Tries to parse the Entity Key from the string to remove the Entity from Datastore
+        Key key;
+        try {
+            key = KeyFactory.stringToKey(id);
```
Comment (alexandrewalsh) [inline src/main/java/com/google/sps/storage/BookmarkStorageManager.java:106]: nit: i think you can just do `Key key = KeyFactory.stringToKey(id)` instead (just saving the one `Key key;` line)

File: src/main/webapp/player.html [hunk 2]
```diff
@@ -75,5 +95,12 @@ <h1 id="resultsHeader" class="text-center hide">Keywords in Video</h1>
                 <div id="output"></div>
             </div>
         </div>
+
+  <div id="bookmark-display-div" class="card-columns">
+  </div>
+
+        <script src="https://code.jquery.com/jquery-3.2.1.slim.min.js" integrity="sha384-KJ3o2DKtIkvYIK3UENzmM7KCkRr/rE9/Qpg6aAZGJwFDMVNA/GpGFF93hXpG5KkN" crossorigin="anonymous"></script>
+        <script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.12.9/umd/popper.min.js" integrity="sha384-ApNbgh9B+Y1QKtv3Rn7W3mgPxhU9K/ScQsAP7hUibX39j7fakFPskvXusvfa0b4Q" crossorigin="anonymous"></script>
+        <script src="https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0/js/bootstrap.min.js" integrity="sha384-JZR6Spejh4U02d8jOt6vLEHfe/JQGiRRSQQxSfFWpi1MquVdAyjUar5+76PVCmYl" crossorigin="anonymous"></script>
```
Comment (alexandrewalsh) [inline src/main/webapp/player.html:104]: what are these?

File: src/main/java/com/google/sps/servlets/BookmarkHandlerServlet.java [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+        List<Bookmark> bookmarks = new ArrayList<>();
+        try {
+            bookmarks = storage.getAllBookmarks(email, videoId);
+        } catch (BookmarkStorageException e) {
+            response.getWriter().println(e.getReasonAsJsonString());
+            return;
+        }
+
+        // Converts the fetched database data and passes it to the front end
+        Gson gson = new Gson();
```
Comment (jiashenwang) [inline src/main/java/com/google/sps/servlets/BookmarkHandlerServlet.java:71]: Nit: Let's move gson to a member var so we do need to init in different functions

File: src/main/webapp/js/captions.js [hunk 4]
```diff
@@ -376,6 +402,132 @@ function setClickableEntities() {
...
+ * @param list - The list of Bookmark objects
+ */
+function displayBookmarks(list) {
+    // Resets the global bookmarks variable to only store current bookmarks
+    bookmarks = {};
+
+    // Builds the HTML text to display on page
+    var output = '';
+    for (bookmark of list) {
+        bookmarks[bookmark.id] = {'timestamp': bookmark.timestamp, 'content': bookmark.content};
```
Comment (jiashenwang) [inline src/main/webapp/js/captions.js:436]: Nit: This function is a little too long. Let's create a function called ```renderBookmark(bookmark)``` to extract the rendering logic out.

File: src/main/webapp/js/captions.js [hunk 5]
```diff
@@ -376,6 +402,132 @@ function setClickableEntities() {
...
+        output += '">Remove</button></div></div>'
+    }
+    
+    // Inserts the HTML text to the page
+    $('#bookmark-display-div').html(output);
+
+    // Removes click listeners from buttons to remove bookmarks to redefine click functionality
+    // Uses a fetch POST request to remove the current bookmark from the database
+    $('.remove-bookmark').off('click');
+    $('.remove-bookmark').click(function() {
```
Comment (jiashenwang) [inline src/main/webapp/js/captions.js:452]: Extract the onclick listener out

File: src/main/webapp/js/captions.js [hunk 6]
```diff
@@ -376,6 +402,132 @@ function setClickableEntities() {
...
+            body: params,
+        }).then((response) => response.json()).then(json => {
+            displayBookmarks(json);
+        });
+    });
+
+    // Removes click listeners from buttons to show bookmark content to redefine click functionality
+    // Toggles between showing and hiding the bookmark content
+    $('.view-bookmark').off('click');
+    $('.view-bookmark').click(function() {
```
Comment (jiashenwang) [inline src/main/webapp/js/captions.js:475]: Ditto

File: src/main/java/com/google/sps/servlets/BookmarkHandlerServlet.java (hunk 3 above)
Comment (yimingnzhao) [reply src/main/java/com/google/sps/servlets/BookmarkHandlerServlet.java]: addressed

File: src/main/webapp/js/captions.js (hunk 4 above)
Comment (yimingnzhao) [reply src/main/webapp/js/captions.js]: will address in `tab-searching` branch, which has more front end designs

File: src/main/webapp/js/captions.js (hunk 5 above)
Comment (yimingnzhao) [reply src/main/webapp/js/captions.js:432]: addressed

File: src/main/webapp/js/captions.js (hunk 6 above)
Comment (yimingnzhao) [reply src/main/webapp/js/captions.js:460]: addressed

**Pull request merged**